### ISSUE TITLE
Sporadic reproducible GetListing, Download and other failures: IOException on control connection

### ISSUE THREAD

**Issue by FanDjango** [robinrodricks/FluentFTP#1030]
This has been reported quite a few times.

### What does the failure look like?

Typically, it manifests as an `IOException`. Depending on the API call being performed, this exception can cause 

- a `GetListing` result to be empty while the exeption is ignored
- other commands to fail with an `IOException`.

If you look closely at the `IOException`, you might see:
`an attempt was made to access a socket in a way forbidden by its access permissions`

### When does it happen?

It happens on an  **control connection** (only if encrypted) when you have issued many API calls (that execute commands and get the response). 

Reported by (ignore the incorrect title): 

@omanThYp https://github.com/robinrodricks/FluentFTP/issues/104#issuecomment-632437591
@oleksabor https://github.com/robinrodricks/FluentFTP/issues/104#issuecomment-953899294

#752 could such a case, but sadly there is no log

See also #723 : one of the "strange" ways this problem can show up.

**Comment by FanDjango** [comment]
Some first results of testing, using @oleksabor initial approach as a basis. 

The goal was to find out a little bit more about the real problem and to remove conjecture.

1. I have modified the test program to **only** do the following (up to 500 times):

```
CWD /home/testuser
PWD
```

These commands **do not use** a data connection. No data connection (SSL) needs to be created and disposed. They only exercise the control connection.

These commands are iterated up to 500 times with no waits in between.

They always fail in the same way **on the 474th iteration**, on all of my available client systems.

2. The test suite will run all 500 iterations without failure if encryption is turned off.

3. The test suite encounters the _problem or multiple problems_ if encryption is turned on.

4. Disabling stream buffering does not help.

5. Disabling or enabling socket polling does not have any effect on the problem occurrence(s).

6. .NET 6.0 or .NET Framework 4.8 targets both exhibit the same problem(s).

7. Any problems occur after the same amount of iterations - deterministic and not random. Independent of host system.

8. The errors occur in `GetReply...ReadLine --> FtpSocketStream.ReadLine --> Read`. I put a try/catch into `FtpSocketStream.Read`: The errors themselves are **always**, (note: also in the original Getlisting/Download szenario) **"an attempt was made to access a socket in a way forbidden by its access permissions"**, an IOException, after which the socket is no longer connected, which is reacted to in different ways when they percolate upwards, depending on the function that you called. In a `GetListing()`, in a `Download...`, depending on the code that catches this exception, the further behaviour difffers. The control connection is closed though, which can cause a reconnect attempt, thus causing confusion if you don`t see that happening, or you can get "empty directories". Standardizing this behaviour could be a todo-item.

9. It looks like continuous reading from the SSL control connection stream in a huge amount of single byte READ operations is somehow exhausting some resource inside the implementation of SslStream.

**Comment by FanDjango** [comment]
I then wrote, in #104 

## 
@oleksabor Using your exact github code and connecting to a very high performace proftpd server of my own, I can reproduce the issue.

### I am getting the same errors ( your sln is using FluentFTP V35).

After upgrading to FluentFTP V42 and making the few neccessary ->V40 migration changes, **the problem STILL can be reproduced**.

So now we can start to search for the problem.

**Comment by FanDjango** [comment]
Some first results of testing, using @oleksabor initial approach as a basis. 

The goal was to find out a little bit more about the real problem and to remove conjecture.

1. I have modified the test program to **only** do the following (up to 500 times):

```
CWD /home/testuser
PWD
```

These commands **do not use** a data connection. No data connection (SSL) needs to be created and disposed. They only exercise the control connection.

These commands are iterated up to 500 times with no waits in between.

They always fail in the same way **on the 474th iteration**, on all of my available client systems.

2. The test suite will run all 500 iterations without failure if encryption is turned off.

3. The test suite encounters the _problem or multiple problems_ if encryption is turned on.

4. Disabling stream buffering does not help.

5. Disabling or enabling socket polling does not have any effect on the problem occurrence(s).

6. .NET 6.0 or .NET Framework 4.8 targets both exhibit the same problem(s).

7. Any problems occur after the same amount of iterations - deterministic and not random. Independent of host system.

8. The errors occur in `GetReply...ReadLine --> FtpSocketStream.ReadLine --> Read`. I put a try/catch into `FtpSocketStream.Read`: The errors themselves are **always**, (note: also in the original Getlisting/Download szenario) **"an attempt was made to access a socket in a way forbidden by its access permissions"**, an IOException, after which the socket is no longer connected, which is reacted to in different ways when they percolate upwards, depending on the function that you called. In a `GetListing()`, in a `Download...`, depending on the code that catches this exception, the further behaviour difffers. The control connection is closed though, which can cause a reconnect attempt, thus causing confusion if you don`t see that happening, or you can get "empty directories". Standardizing this behaviour could be a todo-item.

9. It looks like continuous reading from the SSL control connection stream in a huge amount of single byte READ operations is somehow exhausting some resource inside the implementation of SslStream.

**Comment by FanDjango** [comment]
I have further reduced the tests to
```
Command:  NOOP
Status:   Waiting for response to: NOOP
Response: 200 NOOP command successful
```
I can do 1057 of these commands and then I get the above described failure.

Choosing a "longer" command (i.e. more bytes) does not change the number of commands before the exception occurs.

So I my environment I can cycle about 1057 commands/responses and then I get an IOException.

**Comment by FanDjango** [comment]
Conversely, if I do
```
Command:  PASV
Response: 227 Entering Passive Mode (192,168,1,109,132,201).
Command:  MLSD /home/mike
Response: 150 Opening BINARY mode data connection for MLSD
....listing deleted....
Response: 226 Transfer complete
```
Guess how many? 289.
Looks like we must count the response lines, not the commands.

**Comment by robinrodricks** [comment]
Made a small change to only honor `SslSessionLength` if the FTP connection is using FTPS (IsEncrypted).

**Comment by FanDjango** [comment]
This is now released as V42.1.0
